### PR TITLE
fix: Detect duplicated context values in `distinct values()` + `union()` + `duplicate values()` 

### LINKED ISSUE #771: distinct values() doesn't work for context items
**Describe the bug**
The built-in function [distinct values()](https://docs.camunda.io/docs/components/modeler/feel/builtin-functions/feel-built-in-functions-list/#distinct-valueslist) doesn't filter duplicates for a list of context items.

```
distinct values([{x:1},{x:2},{x:1}])
```

**To Reproduce**
1. Evaluate the expression
```
distinct values([{x:1},{x:2},{x:1}])
```
2. Verify that the result contains `{x:1}` twice 

**Expected behavior**
The function `distinct values()` should filter duplicates for a list of context items.

```
distinct values([{x:1},{x:2},{x:1}])
// [{"x":1},{"x":2}]
```

**Environment**
* FEEL engine version: `1.17.3`
* Affects: 
  * Camunda Automation Platform 7: [7.x] 
  * Zeebe broker: `8.3.3`
* Support case: https://jira.camunda.com/browse/SUPPORT-23030

### LINKED ISSUE #772: union() doesn't work for context items
**Describe the bug**
The built-in function [union](https://docs.camunda.io/docs/components/modeler/feel/builtin-functions/feel-built-in-functions-list/#unionlist) doesn't filter duplicates for a list of context items.

```
union([{x:1},{y:2}],[{x:1}, {z:3}])
```

**To Reproduce**
1. Evaluate the expression
```
union([{x:1},{y:2}],[{x:1}, {z:3}])
```
2. Verify that the result contains `{x:1}` twice 

**Expected behavior**
The function `union()` should filter duplicates for a list of context items.

```
union([{x:1},{y:2}],[{x:1}, {z:3}])
// [{"x":1},{"y":2}, {"z":3}]
```

**Environment**
* FEEL engine version: `1.17.3`
* Affects: 
  * Camunda Automation Platform 7: [7.x] 
  * Zeebe broker: `8.3.3`
* Support case: https://jira.camunda.com/browse/SUPPORT-23030

### LINKED ISSUE #773: duplicate values() doesn't work for context items
**Describe the bug**
The built-in function [duplicate values()](https://docs.camunda.io/docs/components/modeler/feel/builtin-functions/feel-built-in-functions-list/#duplicate-valueslist) doesn't filter duplicates for a list of context items.

```
duplicate values([{x:1},{x:2},{x:1}])
```

**To Reproduce**
1. Evaluate the expression
```
duplicate values([{x:1},{x:2},{x:1}])
```
2. Verify that the result is an empty list 

**Expected behavior**
The function `duplicate values()` should find duplicates for a list of context items.

```
duplicate values([{x:1},{x:2},{x:1}])
// [{"x":1}]
```

**Environment**
* FEEL engine version: `1.17.3`
* Affects: 
  * Camunda Automation Platform 7: [7.x] 
  * Zeebe broker: `8.3.3`

### PR DESCRIPTION
**Pull request by saig0**
## Description

Fixes the detection of duplicated context values in the functions:
- `distinct values()`
- `union()`
- `duplicate values()`

## Related issues

closes #771 
closes #772 
closes #773

### CODE REVIEW

**Comment by saig0** [comment]
@nicpuppa please take a moment to review the changes. Thank you. :cake:

File: src/main/scala/org/camunda/feel/impl/interpreter/ValComparator.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    // values have a different type
+    case _                                                  => ValError(s"Can't compare '$x' with '$y'")
+  }
+
+  private def compare(x: List[Val], y: List[Val]): ValBoolean = {
+    if (x.size != y.size) {
+      ValBoolean(false)
+
+    } else {
+      val itemsAreEqual = x.zip(y).foldRight(true) { case ((x, y), listIsEqual) =>
```
Comment (nicpuppa) [inline src/main/scala/org/camunda/feel/impl/interpreter/ValComparator.scala:56]: 💭 What about using `forall()` on the zipped collection here ? You can remove also the if/else by merging the conditions:
```
ValBoolean(x.size == y.size && x.zip(y).forall(PREDICATE))
```

File: src/main/scala/org/camunda/feel/impl/interpreter/ValComparator.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            case ValBoolean(itemIsEqual) => itemIsEqual
+            case _                       => false
+          }
+        }
+      }
+      ValBoolean(itemsAreEqual)
+    }
+  }
+
+  private def compare(x: Context, y: Context): ValBoolean = {
```
Comment (nicpuppa) [inline src/main/scala/org/camunda/feel/impl/interpreter/ValComparator.scala:68]: 💭 You can use the same approach suggested above also here

**Review by nicpuppa — APPROVED** [review]
In general looks good to me 🚀  

I tried to challenge a bit your solution but I'm not super confident of my Scala knowledge, so if my suggestion are wrong ignore them 😄

File: src/main/scala/org/camunda/feel/impl/interpreter/ValComparator.scala [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+  def equals(x: Val, y: Val): Boolean = compare(x, y) match {
+    case ValBoolean(isEqual) => isEqual
+    case _                   => false
+  }
+
+  def compare(x: Val, y: Val): Val = (x, y) match {
+    // both values are null
+    case (ValNull, _)                                       => ValBoolean(ValNull == y.toOption.getOrElse(ValNull))
+    case (_, ValNull)                                       => ValBoolean(x.toOption.getOrElse(ValNull) == ValNull)
```
[the remark this answers is not in the document]
Comment (saig0) [reply src/main/scala/org/camunda/feel/impl/interpreter/ValComparator.scala:33]: This is a nice idea. :+1: However, for version 1.6, we need to consider that `ValError` should be treated as `ValNull`.

**Comment by saig0** [comment]
@nicpuppa thank you for your fast review and good suggestions. :bow:

**Pull request merged**